### PR TITLE
perf(column): keep track of number of lines that hold up the 'signcolumn'

### PR DESCRIPTION
**Pull request by luukvbaal**
Problem:  The entire marktree needs to be traversed each time a sign is removed from the sentinel line.
Solution: Remove sentinel line and instead keep track of the number of lines that hold up the 'signcolumn' in "max_count". Adjust this number for added/removed signs, and set it to 0 when the maximum number of signs on a line changes. Only when "max_count" is decremented to 0 due to sign removal do we need to check the entire buffer.

Also replace "invalid_top" and "invalid_bot" with a map of invalid ranges, further reducing the number of lines to be checked.

Also improve tree traversal when counting the number of signs. Instead of looping over the to be checked range and counting the overlap for each row, keep track of the overlap in an array and add this to the count.

### CODE REVIEW

**Comment by luukvbaal** [comment]
I've confirmed that indeed the number of times the entire buffer needs to be validated is greatly reduced. We only have to validate the entire buffer twice in `make functionaltest` vs 72 times on `master`.

Despite this, we end up checking **more** lines, because we check the range of the added/removed sign immediately upon sign insertion/deletion. For real usage I would think this PR is already an improvement (`make functionaltest` only tests small buffers). But I wanna see if checking for added/removed signs can be somehow delayed. So that when multiple signs are added to the same range, counting the number of signs on that range is not done multiple times. Which is where I presume this difference in number of checked lines is coming from.

**Comment by luukvbaal** [comment]
Both the number of times the signcolumn is invalidated, and the number of lines checked is now significantly reduced.

File: src/nvim/decoration.c [hunk 1]
```diff
@@ -792,50 +789,95 @@ DecorSignHighlight *decor_find_sign(DecorInline decor)
   }
 }
 
-// Increase the signcolumn size and update the sentinel line if necessary for
-// the invalidated range.
-void decor_validate_signcols(buf_T *buf, int max)
+/// Add overlapping signs in current row to count. If it is greater than
+/// current max, set it and make b_signcols valid. Adjust max_count if
+/// row has or had max signs.
+void buf_signcols_count_row(buf_T *buf, const int overlap[], int idx, int count, bool del)
```
Comment (lewis6991) [inline src/nvim/decoration.c:795]: `overlap` and `idx` are only used once. I would move the count adjustment outside to make this function less convoluted.

File: src/nvim/drawscreen.c [hunk 2]
```diff
@@ -1200,16 +1204,45 @@ void comp_col(void)
   set_vim_var_nr(VV_ECHOSPACE, sc_col - 1);
 }
 
-static void redraw_win_signcol(win_T *wp)
+static int buf_signcols_validate(win_T *wp, buf_T *buf, bool force)
```
Comment (lewis6991) [inline src/nvim/drawscreen.c:1207]: Should this go in the same file as `buf_signcols_validate_range`?

File: src/nvim/buffer_defs.h [hunk 3]
```diff
@@ -703,11 +703,10 @@ struct file_buffer {
...
 
   struct {
-    int size;                   // last calculated number of sign columns
-    int max;                    // maximum value size is valid for.
-    linenr_T sentinel;          // a line number which is holding up the signcolumn
-    linenr_T invalid_top;       // first invalid line number that needs to be checked
-    linenr_T invalid_bot;       // last invalid line number that needs to be checked
+    int max;                    // maximum number of signs on a single line
+    int max_count;              // number of lines with max number of signs
+    bool resized;               // whether max changed at start of redraw
```
Comment (lewis6991) [inline src/nvim/buffer_defs.h:708]: Does this need to be state? It looks like this can be used as the return value for `buf_signcols_validate`, `buf_signcols_validate_range` and `buf_signcols_set_invalid`.

File: src/nvim/extmark.c [hunk 4]
```diff
@@ -67,18 +67,18 @@ void extmark_set(buf_T *buf, uint32_t ns_id, uint32_t *idp, int row, colnr_T col
...
-
-          // not paired: we can revise in place
           mt_itr_rawkey(itr).flags &= (uint16_t) ~MT_FLAG_EXTERNAL_MASK;
           mt_itr_rawkey(itr).flags |= flags;
           mt_itr_rawkey(itr).decor_data = decor.data;
           goto revised;
         }
-        buf_decor_remove(buf, old_mark.pos.row, old_mark.pos.row, mt_decor(old_mark), true);
         marktree_del_itr(buf->b_marktree, itr, false);
+        buf_decor_remove(buf, old_mark.pos.row, old_mark.pos.row, mt_decor(old_mark), true);
```
Comment (lewis6991) [inline src/nvim/extmark.c:81]: Is this change necessary?

File: src/nvim/buffer_defs.h [hunk 5]
```diff
@@ -703,11 +703,10 @@ struct file_buffer {
...
+    Map(int, int) invalid[1];   // map of invalid ranges to be checked
```
[the remark this answers is not in the document]
Comment (luukvbaal) [reply src/nvim/buffer_defs.h:709]: Not sure tbh, to be consistent with `b_extmark_ns`. Presumably just to be able to call `map_` functions with without taking the address.

File: src/nvim/drawscreen.c (hunk 2 above)
Comment (luukvbaal) [reply src/nvim/drawscreen.c:1207]: Sure, both are only used in `drawscreen.c` but have more to do with decorations I guess. Would `static` `drawscreen.c` functions be preferred or exported in `decorations.c`?

File: src/nvim/buffer_defs.h (hunk 3 above)
Comment (luukvbaal) [reply src/nvim/buffer_defs.h:708]: Sure, it's only used to determine whether 'statuscolumn' needs to be rebuild when 'signcolumn' is "number/no". Because `w_scwidth` is and stays 0 then.

File: src/nvim/drawscreen.c (hunk 2 above)
Comment (lewis6991) [reply src/nvim/drawscreen.c:1207]: I think not `drawscreen.c`. I considered adding a new file just for column/signs, but if not that then maybe `decoration.c` for now.

File: src/nvim/extmark.c (hunk 4 above)
Comment (luukvbaal) [reply src/nvim/extmark.c:81]: Hmm since the counting is now delayed to the start of a redraw it no longer is. Can revert I guess but I think it's more correct in this order in case `buf_decor_remove()` ever handles other side effects which expect the mark to already be removed from the marktree.

Comment (lewis6991) [reply src/nvim/extmark.c:81]: The other line only deletes an iterator. It doesn't change the marktree.

Comment (luukvbaal) [reply src/nvim/extmark.c:81]: Uhm I don't think that's true. `marktree_del_itr()` removes the key pointed to by the iterator?

File: src/nvim/buffer_defs.h (hunk 3 above)
Comment (luukvbaal) [reply src/nvim/buffer_defs.h:708]: Is it necessary to remove? It's more convenient as a globally accessible variable.

Comment (lewis6991) [reply src/nvim/buffer_defs.h:708]: Global state should generally be avoided, especially in a repository with a regular rotation of contributors who don't always 100% understand the code they have inherited. There are plenty of resources that explain this (quick one from google: https://softwareengineering.stackexchange.com/questions/148108/why-is-global-state-so-evil).

I know this isn't exactly global state, but given the size of the buf_T struct, the same principles all apply.

Sometimes the benefits of convenience can justify it, time is a real constraint after all, but I don't think that's the case here. I think removing this piece of state will make the code cleaner, even though it is a little more effort.

Comment (lewis6991) [reply src/nvim/buffer_defs.h:708]: Ok, so in this case it does actually need to be state.

File: src/nvim/drawscreen.c [hunk 6]
```diff
@@ -619,6 +621,8 @@ int update_screen(void)
   screen_search_hl.rm.regprog = NULL;
 
   FOR_ALL_WINDOWS_IN_TAB(wp, curtab) {
+    wp->w_buffer->b_signcols.resized = false;
```
[the remark this answers is not in the document]
Comment (luukvbaal) [reply src/nvim/drawscreen.c:624]: Actually it needs to stay `true` until the above loop has ended. Otherwise, subsequent encountered windows displaying the same buffer will not have their statuscolumn rebuild. I.e. this would fail:
https://github.com/neovim/neovim/blob/ca7f8786a0eb578895400e23cd21e25cc0f91800/test/functional/ui/statuscolumn_spec.lua#L838-L845

Could add a nested loop over all windows in `buf_signcols_validate()` and move it there to mark all windows holding the validated buffer.

**Comment by luukvbaal** [comment]
> If we want to keep the invalid map I think it needs to also store how many signs were removed from a line.

I've gone ahead and done this, unfortunately making the diff slightly bigger. But like I said without adding the invalid map (just adding `b_signcols.max_count`), this PR actually increases the number of lines checked. Probably because the current `b_signcols.sentinel` is disproportionally effective in `make functionaltest`. Even so, now at least we can be sure it is a pure performance improvement.

File: src/nvim/decoration.c [hunk 7]
```diff
@@ -792,50 +792,113 @@ DecorSignHighlight *decor_find_sign(DecorInline decor)
   }
 }
 
-// Increase the signcolumn size and update the sentinel line if necessary for
-// the invalidated range.
-void decor_validate_signcols(buf_T *buf, int max)
+/// If "count" is greater than current max, set it and make b_signcols valid.
+/// Adjust max_count if validated row has or had max signs.
+void buf_signcols_validate_row(buf_T *buf, int count, int add)
```
Comment (lewis6991) [inline src/nvim/decoration.c:797]:
```suggestion
static void buf_signcols_validate_row(buf_T *buf, int count, int add)
```

File: src/nvim/decoration.c [hunk 8]
```diff
@@ -792,50 +792,113 @@ DecorSignHighlight *decor_find_sign(DecorInline decor)
...
+  int del = add < 0 ? -add : 0;
+  if (count > buf->b_signcols.max) {
+    buf->b_signcols.max = count;
+    buf->b_signcols.resized = true;
+    buf->b_signcols.max_count = 0;
+  }
+  buf->b_signcols.max_count += (count == buf->b_signcols.max - del) * (del ? -1 : 1);
+}
+
+void buf_signcols_validate_range(buf_T *buf, int row1, int row2, int add)
```
Comment (lewis6991) [inline src/nvim/decoration.c:808]:
```suggestion
static void buf_signcols_validate_range(buf_T *buf, int row1, int row2, int add)
```

File: src/nvim/decoration.c [hunk 9]
```diff
@@ -792,50 +792,113 @@ DecorSignHighlight *decor_find_sign(DecorInline decor)
...
+      buf->b_signcols.max += range.add;
+      buf_signcols_validate_range(buf, 0, buf->b_ml.ml_line_count, 0);
+      break;
     }
+  });
+  map_clear(int, buf->b_signcols.invalid);
+  return buf->b_signcols.max;
+}
+
+void buf_signcols_invalidate_range(buf_T *buf, int row1, int row2, int add)
```
Comment (lewis6991) [inline src/nvim/decoration.c:872]:
```suggestion
static void buf_signcols_invalidate_range(buf_T *buf, int row1, int row2, int add)
```

File: src/nvim/decoration.c [hunk 10]
```diff
@@ -792,50 +792,113 @@ DecorSignHighlight *decor_find_sign(DecorInline decor)
...
-    MarkTreeIter itr[1];
-    if (!marktree_itr_get_overlap(buf->b_marktree, currow, 0, itr)) {
-      continue;
+  int del = add < 0 ? -add : 0;
+  if (count > buf->b_signcols.max) {
+    buf->b_signcols.max = count;
+    buf->b_signcols.resized = true;
+    buf->b_signcols.max_count = 0;
+  }
+  buf->b_signcols.max_count += (count == buf->b_signcols.max - del) * (del ? -1 : 1);
```
Comment (lewis6991) [inline src/nvim/decoration.c:805]: I don't think it's good practice to rely on the fact that `==` returns `0/1`.

Comment (lewis6991) [inline src/nvim/decoration.c:805]:
```suggestion
  buf->b_signcols.max_count += ((count == buf->b_signcols.max - del) ? 1 : 0) * (del ? -1 : 1);
```

Comment (lewis6991) [inline src/nvim/decoration.c:805]:
```suggestion
  if (count -- buf->b_signcols.max - del) {
    buf->b_signcols.max_count += del ? -1 : 1;
  }
```

File: src/nvim/decoration.c [hunk 11]
```diff
@@ -792,50 +792,113 @@ DecorSignHighlight *decor_find_sign(DecorInline decor)
...
+  xfree(overlap);
+}
+
+int buf_signcols_validate(win_T *wp, buf_T *buf, bool force)
+{
+  int start;
+  SignRange range;
+  map_foreach(buf->b_signcols.invalid, start, range, {
+    if ((!force || buf->b_signcols.resized)
+        && range.add > 0 && buf->b_signcols.max >= wp->w_maxscwidth) {
```
Comment (lewis6991) [inline src/nvim/decoration.c:857]: It's not obvious to me what this condition means. Can we add a short comment?

File: src/nvim/decoration.c [hunk 12]
```diff
@@ -792,50 +792,113 @@ DecorSignHighlight *decor_find_sign(DecorInline decor)
...
+void buf_signcols_validate_range(buf_T *buf, int row1, int row2, int add)
+{
+  int count = 0;
+  int currow = row1;
+  MTPair pair = { 0 };
+  MarkTreeIter itr[1];
+  int *overlap = xcalloc(sizeof(int), (size_t)(row2 + 1 - row1));
+
+  marktree_itr_get_overlap(buf->b_marktree, currow, 0, itr);
+  while (marktree_itr_step_overlap(buf->b_marktree, itr, &pair)) {
```
Comment (lewis6991) [inline src/nvim/decoration.c:817]: I know this might be a common way of finding range intersections, but the API for extmarks here is a bit esoteric, so some descriptions of what is happening will make this easier to follow. I.e. why is this only done for `row1`?

File: src/nvim/drawscreen.c [hunk 13]
```diff
@@ -1200,16 +1192,34 @@ void comp_col(void)
...
+    width = buf_signcols_validate(wp, buf, false);
+  }
+
   int scwidth = wp->w_scwidth;
-  wp->w_scwidth = win_signcol_count(wp);
-  if (wp->w_scwidth != scwidth) {
+  wp->w_scwidth = MAX(wp->w_minscwidth, width);
+  if (scwidth != wp->w_scwidth || rebuild_stc) {
+    wp->w_lines_valid = 0;
+    wp->w_redr_type = UPD_NOT_VALID;
```
[the remark this answers is not in the document]
Comment (luukvbaal) [reply src/nvim/drawscreen.c:1221]: It does but this is called from `update_screen()` in which directly modifying `w_redr_type` is more idiomatic.

File: src/nvim/decoration.c (hunk 12 above)
Comment (luukvbaal) [reply src/nvim/decoration.c:817]: The explanation is given in the `marktree_get_overlap()` docstring. Can add a comment referring to that I guess.

File: src/nvim/drawscreen.c (hunk 13 above)
[the remark this answers is not in the document]
Comment (lewis6991) [reply src/nvim/drawscreen.c:1221]: But this is `win_redraw_signcols()`, the caller shouldn't matter. I don't think it is reasonable for the reader to have deep knowledge of the call stack in order to understand the code. 

This should probably return a boolean then you can set `w_redr_type` in `update_screen()` directly.

File: src/nvim/decoration.c [hunk 14]
```diff
@@ -792,50 +792,113 @@ DecorSignHighlight *decor_find_sign(DecorInline decor)
...
+    buf->b_signcols.max = count;
+    buf->b_signcols.resized = true;
+    buf->b_signcols.max_count = 0;
+  }
+  buf->b_signcols.max_count += (count == buf->b_signcols.max - del) * (del ? -1 : 1);
+}
+
+void buf_signcols_validate_range(buf_T *buf, int row1, int row2, int add)
+{
+  int count = 0;
```
Comment (lewis6991) [inline src/nvim/decoration.c:810]: What are we counting?

File: src/nvim/decoration.c [hunk 15]
```diff
@@ -792,50 +792,128 @@ DecorSignHighlight *decor_find_sign(DecorInline decor)
...
-      MTKey mark = marktree_itr_current(itr);
-      if (mark.pos.row != currow) {
-        break;
-      }
-      if (!mt_invalid(mark) && !mt_end(mark) && (mark.flags & MT_FLAG_DECOR_SIGNTEXT)) {
-        count++;
-      }
-      marktree_itr_next(buf->b_marktree, itr);
+  // Continue traversing the marktree until beyond "row2". Increment "count" for
+  // the start of a mark, increment the overlap array until the end of a paired mark.
```
Comment (lewis6991) [inline src/nvim/decoration.c:832]: Perfect

**Review by lewis6991 — APPROVED** [review]
This looks great now. Thanks.

**Comment by luukvbaal** [comment]
Thanks for the review!
